### ISSUE TITLE
syncthing: Syncthing fails to setup inotify

### ISSUE THREAD

**Issue by alexanderharm** [SynoCommunity/spksrc#3540]
_For new Package Requests, see the guidelines_  

### Setup
syncthing
0.14.49-15 (self-updated to 0.14.54)

Synology 1517+
x86_64
DSM 6.2.1-23824 Update 1

### Expected behavior
Watching folders should work.

### Actual behavior
Setting up inotify watches fails with permissions error.

### Steps to reproduce
_1._ Install syncthing
_2._ Enable filesystem watching
_3._ Will fail

#### Package log

```
2018-12-06 00:34:11 folder.go:611 basic /volume1/KUNDEN_1 Watch . Matcher/[$RECYCLE.BIN **/$RECYCLE.BIN $RECYCLE.BIN/** **/$RECYCLE.BIN/** $Recycle.Bin **/$Recycle.Bin $Recycle.Bin/** **/$Recycle.Bin/** (?d)lost+found (?d)**/lost+found (?d)lost+found/** (?d)**/lost+found/** (?d).DocumentRevisions-V100 (?d)**/.DocumentRevisions-V100 (?d).DocumentRevisions-V100/** (?d)**/.DocumentRevisions-V100/** (?d).TemporaryItems (?d)**/.TemporaryItems (?d).TemporaryItems/** (?d)**/.TemporaryItems/** (?d).fseventsd (?d)**/.fseventsd (?d).fseventsd/** (?d)**/.fseventsd/** (?d).iCloud (?d)**/.iCloud (?d).iCloud/** (?d)**/.iCloud/** (?d).DS_Store (?d)**/.DS_Store (?d).DS_Store/** (?d)**/.DS_Store/** (?d).Spotlight-V100 (?d)**/.Spotlight-V100 (?d).Spotlight-V100/** (?d)**/.Spotlight-V100/** (?d).Trashes (?d)**/.Trashes (?d).Trashes/** (?d)**/.Trashes/** .Trash-* **/.Trash-* .Trash-*/** **/.Trash-*/** ehthumbs.db **/ehthumbs.db ehthumbs.db/** **/ehthumbs.db/** (?d)desktop.ini (?d)**/desktop.ini (?d)desktop.ini/** (?d)**/desktop.ini/** (?d)Thumbs.db (?d)**/Thumbs.db (?d)Thumbs.db/** (?d)**/Thumbs.db/** (?d)~* (?d)**/~* (?d)~*/** (?d)**/~*/** *~ **/*~ *~/** **/*~/** .~lock.* **/.~lock.* .~lock.*/** **/.~lock.*/** *.part **/*.part *.part/** **/*.part/** *.filepart **/*.filepart *.filepart/** **/*.filepart/** .csync_journal.db **/.csync_journal.db .csync_journal.db/** **/.csync_journal.db/** .csync_journal.db.tmp **/.csync_journal.db.tmp .csync_journal.db.tmp/** **/.csync_journal.db.tmp/** *.swn **/*.swn *.swn/** **/*.swn/** *.swp **/*.swp *.swp/** **/*.swp/** *.swo **/*.swo *.swo/** **/*.swo/** .dropbox **/.dropbox .dropbox/** **/.dropbox/** .dropbox.attr **/.dropbox.attr .dropbox.attr/** **/.dropbox.attr/** *.crdownload **/*.crdownload *.crdownload/** **/*.crdownload/** (?d)@eaDir (?d)**/@eaDir (?d)@eaDir/** (?d)**/@eaDir/** @SynoResource **/@SynoResource @SynoResource/** **/@SynoResource/** .@__thumb **/.@__thumb .@__thumb/** **/.@__thumb/** .thumbnails **/.thumbnails .thumbnails/** **/.thumbnails/** ._* **/._* ._*/** **/._*/** #snapshot **/#snapshot #snapshot/** **/#snapshot/** (?d)node_modules (?d)**/node_modules (?d)node_modules/** (?d)**/node_modules/** #recycle **/#recycle #recycle/** **/#recycle/** /Autofleetcontrol /Autofleetcontrol/** /DCBK /DCBK/** /GRUNDLAGEN?WBT /GRUNDLAGEN?WBT/** /Grasdorf /Grasdorf/** /KunstAkademie_Foerderverein /KunstAkademie_Foerderverein/**]@0xc0002481b0 true error while traversing /volume1/KUNDEN_1/_Schriften_für_alle: permission denied
2018-12-06 00:34:11 log 69 FolderWatchStateChanged map[folder:my-folder-id from:error while traversing /volume1/KUNDEN_1/_Schriften_für_alle: permission denied to:error while traversing /volume1/KUNDEN_1/_Schriften_für_alle: permission denied]
2018-12-06 00:34:11 Repeat error while trying to start filesystem watcher for folder "DCBK_SALT_KUNDEN_1" (my-folder-id), trying again in 1min: error while traversing /volume1/KUNDEN_1/_Schriften_für_alle: permission denied
2018-12-06 00:34:11 http: GET "/rest/events?since=68": status 200, 380 bytes in 59050.45 ms
2018-12-06 00:34:12 my-folder-id WithNeedTruncated(7777777-777777N-7777777-777777N-7777777-777777N-7777777-77777Q4)
2018-12-06 00:34:12 progress emitter: bytes completed for my-folder-id: 0
2018-12-06 00:34:12 model@0xc00002e2c0 NeedSize("my-folder-id"): {0 0 0 0 0 0 [] 0}
2018-12-06 00:34:12 folderconfiguration.go:151 basic /volume1/KUNDEN_1 Stat . {0xc005aff520} <nil>
2018-12-06 00:34:12 folderconfiguration.go:169 basic /volume1/KUNDEN_1 Stat .stfolder {0xc005aff5f0} <nil>
2018-12-06 00:34:12 ignore.go:544 basic /volume1/KUNDEN_1 Stat .stignore {0xc005aff6c0} <nil>
2018-12-06 00:34:12 log 70 FolderSummary map[folder:my-folder-id summary:map[watchError:error while traversing /volume1/KUNDEN_1/_Schriften_für_alle: permission denied globalBytes:2355058942912 localDeleted:35807 needSymlinks:0 needDeletes:0 inSyncBytes:2355058942912 ignorePatterns:true errors:0 pullErrors:0 invalid: globalDirectories:40099 localDirectories:40099 stateChanged:2018-12-06 00:26:35.391486179 +0100 CET m=+474.192771854 version:923442 globalFiles:215233 globalDeleted:35829 localSymlinks:6 localBytes:2355058942912 needFiles:0 state:idle globalSymlinks:6 localFiles:215233 needDirectories:0 needBytes:0 inSyncFiles:215233 sequence:923442]]
2018-12-06 00:34:12 stats.DeviceStatisticsReference.WasSeen: REMOTE-NAS-ID
2018-12-06 00:34:12 http: GET "/rest/events?since=69": status 200, 1080 bytes in 1063.83 ms
2018-12-06 00:34:12 my-folder-id WithNeedTruncated(REMOTE-NAS-ID)
2018-12-06 00:34:13 model@0xc00002e2c0 Completion(REMOTE-NAS-ID, "my-folder-id"): 100.000000 (0 / 2355058942912 = 0.000000)
2018-12-06 00:34:13 log 71 FolderCompletion map[device:REMOTE-NAS-ID completion:100 needBytes:0 needItems:0 globalBytes:2355058942912 needDeletes:0 folder:my-folder-id]
```

**Comment by alexanderharm** [comment]
Issue seems to be that the shared folder is also a Synology shared folder. Creating a folder one level below seems to work fine.

**Comment by BenjV** [comment]
In that case the permissions of that Synology shared folder were not set correct by yourself I guess.
You have to give the group sc-syncthing read/write permissions on that share.

**Comment by BenjV** [comment]
Then syncthing is a stupid system which is explicitly looking for those permission setting instead of just trying.
An ACL setting will override  the Posix permissions if present and Synology is migration to ACL's, so if you have given the group sc-syncthing permission to that share syncthing will have access.

Maybe better yo are better off using the Synology Drive package in stead of Syncthing.

**Comment by alexanderharm** [comment]
Maybe... but I have very bad experience using CloudStation/Drive when it comes to large folders. Synology recommends no more than 200.000 files for a quad core Atom. Now I sync 1.500.000 files using Syncthing without any issues. CloudStation was simply unusable... I'm not convinced that it improved that much in the last months.

**Comment by alexanderharm** [comment]
Syncthing devs say that the kernel uses POSIX regardless of ACLs. I'm not expert enough to judge this...

**Comment by BenjV** [comment]
Stupid arrogant developers.
They should not do such a thing, but just try to use the folder.
Permissions should be kept outside an application, especially I you pretend to be an application useable on lots of platforms.

In the past I myself first started to use bittorrent sync but they went commercial and demanded money.
I switched to Syncthing and never have used a more worthless  application which used enormous amount of resources and kept crashing and no wonder because it is written in Go.
Now I am using Drive with functions fine.

One thing I know for sure Drive is a lot better then Syncthing.

**Comment by alexanderharm** [comment]
OK, I investigated a bit further and tried to setup `inotify` watches in the shell:

```
# succeeds
sudo -u sc-syncthing /opt/bin/inotifywait /volume1/KUNDEN_1/
Setting up watches.
Watches established.

# fails
sudo -u sc-syncthing /opt/bin/inotifywait /volume1/KUNDEN_1/_Schriften_für_alle/
Setting up watches.
Couldn't watch /volume1/KUNDEN_1/_Schriften_für_alle/: Permission denied
```

The permissions are as follows:

```
# succeeds
d---------+  1 root  root       606 Dec  6 21:32 KUNDEN_1
[1] group:sc-syncthing:allow:rwxpdDaARWc--:fd--  (level:0)

# fails
d---------+ 1 sadmin       users            826 Feb 27  2018 _Schriften_für_alle
[1] group:sc-syncthing:allow:rwxpdDaARWc--:fd--  (level:1)
```

I tested some more folders:

```
# fails
d---------+ 1 sadmin       users            388 Feb 27  2018 KLEBER
[1] group:sc-syncthing:allow:rwxpdDaARWc--:fd--  (level:1)
# succeeds after
chown sc-syncthing /volume1/KUNDEN_1/KLEBER/

# succeeds
d---------+ 1 sadmin       users            234 Mar  5  2018 _StockPhoto_Datenbank
[1] group:sc-syncthing:allow:rwxpdDaARWc--:fd--  (level:1)

# fails
d---------+ 1 sadmin       users            218 Feb  5  2018 BFGoodrich
[1] group:sc-syncthing:allow:rwxpdDaARWc--:fd--  (level:1)
```

I fail to see any logic in this...

**Comment by m4tt075** [comment]
What is the output of the following commands?
`synoacltool -get "/volume1/KUNDEN_1"`
`synoacltool -get "/volume1/KUNDEN_1/_Schriften_für_alle"`
...and analogous for the other folders that succeed or fail?

**Comment by m4tt075** [comment]
@alexanderharm ^^ We might just have cross-posted. Please report the results for the commands I've suggested and let's see whether we see correlations there...

**Comment by alexanderharm** [comment]
@m4tt075 I posted the the ACLs for the `sc-syncthing` but here is the complete list

```
# /volume1/KUNDEN_1
synoacltool -get /volume1/KUNDEN_1/
ACL version: 1
Archive: has_ACL,is_support_ACL
Owner: [root(user)]
---------------------
	 [0] group::allow:rwxpdDaARWc--:fd--  (level:0)
	 [1] group:sc-syncthing:allow:rwxpdDaARWc--:fd--  (level:0)
	 [2] group:KUNDEN_1:allow:rwxpdDaARWc--:fd--  (level:0)
	 [3] group::allow:rwxpdDaARWc--:fd--  (level:0)
	 [4] group:administrators:allow:rwxpdDaARWc--:fd--  (level:0)
	 [5] group::allow:rwxpdDaARWc--:fd--  (level:0)

# /volume1/KUNDEN_1/_Schriften_für_alle
synoacltool -get /volume1/KUNDEN_1/_Schriften_für_alle/
ACL version: 1
Archive: is_inherit,is_support_ACL
Owner: [sadmin(user)]
---------------------
	 [0] group::allow:rwxpdDaARWc--:fd--  (level:1)
	 [1] group:sc-syncthing:allow:rwxpdDaARWc--:fd--  (level:1)
	 [2] group:KUNDEN_1:allow:rwxpdDaARWc--:fd--  (level:1)
	 [3] group::allow:rwxpdDaARWc--:fd--  (level:1)
	 [4] group:administrators:allow:rwxpdDaARWc--:fd--  (level:1)
	 [5] group::allow:rwxpdDaARWc--:fd--  (level:1)
```

Right now I'm in the process of removing all ACLs with no name and the ones of sc-syncthing. I will then add sc-syncthing again.

**Comment by alexanderharm** [comment]
I doubt it, two more examples here. One fails without fancy characters.

```
# succeeds
synoacltool -get /volume1/KUNDEN_1/_StockPhoto_Datenbank/
ACL version: 1
Archive: is_inherit,is_support_ACL
Owner: [sadmin(user)]
---------------------
	 [0] group::allow:rwxpdDaARWc--:fd--  (level:1)
	 [1] group:sc-syncthing:allow:rwxpdDaARWc--:fd--  (level:1)
	 [2] group:KUNDEN_1:allow:rwxpdDaARWc--:fd--  (level:1)
	 [3] group::allow:rwxpdDaARWc--:fd--  (level:1)
	 [4] group:administrators:allow:rwxpdDaARWc--:fd--  (level:1)
	 [5] group::allow:rwxpdDaARWc--:fd--  (level:1)

# fails
synoacltool -get /volume1/KUNDEN_1/BFGoodrich/
ACL version: 1
Archive: is_inherit,is_support_ACL
Owner: [sadmin(user)]
---------------------
	 [0] group::allow:rwxpdDaARWc--:fd--  (level:1)
	 [1] group:sc-syncthing:allow:rwxpdDaARWc--:fd--  (level:1)
	 [2] group:KUNDEN_1:allow:rwxpdDaARWc--:fd--  (level:1)
	 [3] group::allow:rwxpdDaARWc--:fd--  (level:1)
	 [4] group:administrators:allow:rwxpdDaARWc--:fd--  (level:1)
	 [5] group::allow:rwxpdDaARWc--:fd--  (level:1)
```

**Comment by m4tt075** [comment]
Right. I agree with your approach. Would also try to remove the acls and see what happens then. If I remember correctly there is a special synoacl command to do that. If it works, @BenjV is right though!  ;-)

**Comment by m4tt075** [comment]
Yep, just checked because I could not remember: `synoacltool -del "/Folder/Name"` should do the job.

**Comment by alexanderharm** [comment]
@m4tt075 You are saying CloudStation can handle 18TB of data and 1.5 million files just fine? Not my experience. I also have other customers that run it on a lot less files and it does simply not work reliably... Apart from having a master/slave approach and not a cluster-one like Syncthing...

**Comment by alexanderharm** [comment]
I deleted ACLs and re-added them. Result remains the same. I'm running out of ideas... I will contact Synology's support.

**Comment by m4tt075** [comment]
@alexanderharm Nope, but that the approach developers have been taking is "suboptimal", to put it mildly...  :-)

Re your latest attempt: I would not readd ACL permissions. Just remove those and chown and chmod the folders. What happens then?

**Comment by alexanderharm** [comment]
I'm not sure which approach we are talking about...  but let's put this aside.

Removing ACLs completely does the trick. No chown/chmod necessary... (interesting because the user should not have access, should he?)

**Comment by m4tt075** [comment]
Very interesting indeed. Would like to know what the Synology support has to say about that... OK, at least it seems to work now.

**Comment by alexanderharm** [comment]
Sorry, tested the wrong path :-)

1. Removing ACLs completely doesn't change a thing.
2. Running `chown` does the trick.
3. `chmod` is not necessary, even though 

I will ask Synology, because I fail to see any logic in this.

**Comment by m4tt075** [comment]
Running chown AFTER removing the ACLs would make sense.
Running chown WITHOUT removing the ACLs would not make sense.
What was it?

**Comment by alexanderharm** [comment]
Both. It is only relevant to run `chown` on folders that do not work. There are however folders that do work regardless of `chown` (`/volume1/KUNDEN_1` and `/volume1/KUNDEN_1/_StockPhoto_Datenbank`) with ACLs or without ACLs. If it wouldn't be in my Shell history I wouldn't believe this to be honest.

**Comment by m4tt075** [comment]
Agreed. And that means @BenjV WAS right:
> St**** ****** [syncthing] developers.
> They should not do such a thing, but just try to use the folder.
I don't think this is a Synology problem.

**Comment by alexanderharm** [comment]
Examples:

```
synoacltool -get KLEBER/
ACL version: 1
Archive: is_inherit,is_support_ACL
Owner: [sadmin(user)]
---------------------
	 [0] group:administrators:allow:rwxpdDaARWc--:fd--  (level:1)
	 [1] group:sc-syncthing:allow:rwxpdDaARWc--:fd--  (level:1)
	 [2] group:KUNDEN_1:allow:rwxpdDaARWc--:fd--  (level:1)

ls -l
d---------+ 1 sadmin       users     388 Feb 27  2018 KLEBER

sudo -u sc-syncthing /opt/bin/inotifywait /volume1/KUNDEN_1/KLEBER/
Setting up watches.
Couldn't watch /volume1/KUNDEN_1/KLEBER/: Permission denied

chgrp sc-syncthing KLEBER

sudo -u sc-syncthing /opt/bin/inotifywait /volume1/KUNDEN_1/KLEBER/
Setting up watches.
Couldn't watch /volume1/KUNDEN_1/KLEBER/: Permission denied

chown sc-syncthing KLEBER
sudo -u sc-syncthing /opt/bin/inotifywait /volume1/KUNDEN_1/KLEBER/
Setting up watches.
Watches established.
```

**Comment by alexanderharm** [comment]
I'm testing this with Entware-ng's `inotify-tools`. So this has nothing to do with `Syncthing`!

**Comment by alexanderharm** [comment]
`inotifywait` levers the exact same result as `Syncthing` in repect to the folders that fail. And therefore I believe that part of what the Syncthing devs say is correct. Somehow the POSIX ownership matters when it comes to `inotify` but obviously not always...

**Comment by m4tt075** [comment]
Right, `inotify`, not `syncthing`. If `chown` matters no mattter what the ACL permissions say, the devs obviously are correct. I agree with you. But that is not how permission management is supposed to work on a Synology system. I believe there are two options left:
1.  Either this is a problem with the way how `inotifywait` "evaluates" permissions or...
2. ... it is a problem with the implementation of ACL permissions by Synology.

I suspected option 1. Do you suspect option 2 or what do you mean with "`inotifywait` levers the exact same results as `Syncthing`"?

**Comment by alexanderharm** [comment]
I meant it gives the same result as... meaning the permission errors reported by `Syncthing` are identical to the permission errors reported by `inotifywait`.

AFAIK Synology uses a modified version of `inotify`. They would probably say "enhanced" version. So I suspect there is an incompatibility between standard Linux inotify tools and Synology's implementation. Since there are a couple of posts suggesting that users at least ran `Syncthing` with Synology in the past just fine this incompatibility might have been introduced only recently.

I opened a ticket with Synology. Let's see what they have to say about it.

**Comment by m4tt075** [comment]
Agreed. Thanks for your dilligence and follow-up. Very interested in learning what they'll say.

**Comment by BenjV** [comment]
I think you are on the wrong path.

The problem is that they check the Posix permissions setting instead of just trying to write to the folder.
So by checking the Posix permissions they wrongfully assume the permission is not there, but that permission is given via an ACL which they don't check.
An application should just try to write instead of check the permissions.

That's why the mantra of python developers( and may others) is:
It's better to ask forgiveness than permission

**Comment by publicarray** [comment]
@BenjV 
> kept crashing and no wonder because it is written in Go.

Auch! Anti-Go much.

> An application should just try to write instead of check the permissions.
>
> It's better to ask forgiveness than permission

I can agree when it comes to writing files, but this mind set is dangerous when for example deleting files. This mindset is what damages brand names. Think about the many times Mozilla has implemented things and needed to ask for forgiveness rather than just ask users if they wanted Pocket integration, the new tab page, cliqz experiment etc.

**Comment by alexanderharm** [comment]
@BenjV As mentioned above, I believe there is very little evidence to support your claim. `inotify-tools` which are neither developed by `Syncthing` devs nor written in `go` nor have anything to do with `Syncthing` at all show exactly the same symptoms. Also I think that we should appreciate the time and effort open source developers put into their projects free of charge. I for one am glad that there is a FOSS alternative to commercial solutions. Sure, their software is not bug free but neither is Synology's nor anyone else's.

**Comment by BenjV** [comment]
I never stated that inotify-tools are written in go, I really don't know which tools they used.
I just stated that the only conclusion of your testing is, that both applications only test the Posix permissions and then conclude they permission is incorrect.
That is not the correct way to do it especially an application that should be used with multiple platforms.

And of course you may believe anything you want, but it is never wise to ignore things simply because you don't like them.

@publicarray
Go is a very tiny language (like C), so everything has to be coded by the programmers, no extensive library of bugfree code like for example python so like a program written in C is has to be tested extensively to get all the bug out of the hardly used parts of the code with strange use cases.
Real threading is not available, just some sort of surrogate called channels.
It compiles into one big binary, which seems an advantage but is not at all.
One line of Go code creates a binary of 2MB to begin with.
They dont' use dynamic linking or linked libraries or reuseable code so if you have multiple Go applications everything needed is multiplied in memory.

**Comment by BenjV** [comment]
It is of course impossible to get any difference between accessing "/volume1/KUNDEN_1/A" and "/volume1/KUNDEN_1/B" if both ACL's and Posix permissions are exactly the same.
There is no third (hidden) permission system to my knowledge

Only the application itself can make a difference between those name in this case.

**Comment by alexanderharm** [comment]
This is the answer of the Synology support and it comes at no surprise:

>There is no official inotify-API on our devices. Seems like you have compiled the inotify-tool by yourself (https://github.com/rvoicilas/inotify-tools).
This is not supported so we cannot guarantee any function.

I will leave it as it is and simply set ownership via Task Manager on dirs every night. Then there are no issues. At least none I know of at the moment :-)

Thank you all for your help, patience and input!

**Comment by ymartin59** [comment]
According to investigations (I have not read all), I wonder if partial conclusion would be to "improve" (or "fix") Syncthing source code so that it behaves properly in DSM context - and I am quite sure it is possible to reproduce troubles with test cases on a standard Linux when facl is used on ext4. What is your opinion ?

**Comment by m4tt075** [comment]
Yes. I only believe that they might be as interested in exploring this further as the Synology support was in their response to @alexanderharm, meaning "not at all"...  ;-)
Somebody would have to explore it upstream. @Safihre, not sure you are still interested in this here, but if you are, any thoughts and/or interest from you?

**Comment by BenjV** [comment]
They support the Windows environment and the only permissions system is the use of ACL's in Windows, but according to their comment the world exists of Posix and Windows.

As I said a year or so ago I tried Syncthing for a few months but it was full of bugs and did not perform at all and got stuck very often.
I gave up on trying to use sync functionality and switched to backup, until I installed a few moths ago the Synology Drive which works fine for me at the moment.
For my development environment is sync a much better solution then backup.
